ext/web/compression.rs

```diff
@@ -0,0 +1,187 @@
+
+pub struct GzipCompressorResource {
+  encoder: AsyncRefCell<GzipEncoder<DuplexStream>>,
+  tx: AsyncRefCell<DuplexStream>,
+}
+impl Resource for GzipCompressorResource {
+  fn name(&self) -> Cow<str> {
+    "gzipCompressor".into()
+  }
+
+  fn read(self: Rc<Self>, mut buf: ZeroCopyBuf) -> AsyncResult<usize> {
+    Box::pin(async move {
+      let resource = deno_core::RcRef::map(self, |r| &r.tx);
+      let mut tx = resource.borrow_mut().await;
+      let n = tx.read(&mut buf).await?;
+      Ok(n)
+    })
+  }
+
+  fn write(self: Rc<Self>, buf: ZeroCopyBuf) -> AsyncResult<usize> {
+    Box::pin(async move {
+      let resource = deno_core::RcRef::map(self, |r| &r.encoder);
+      let mut encoder = resource.borrow_mut().await;
+      let n = encoder.write(&buf).await?;
+      Ok(n)
+    })
+  }
+}
+
+pub struct GzipDecompressorResource {
+  decoder: AsyncRefCell<GzipDecoder<DuplexStream>>,
+  tx: AsyncRefCell<DuplexStream>,
+}
+impl Resource for GzipDecompressorResource {
+  fn name(&self) -> Cow<str> {
+    "gzipDecompressor".into()
+  }
+
+  fn read(self: Rc<Self>, mut buf: ZeroCopyBuf) -> AsyncResult<usize> {
+    Box::pin(async move {
+      let resource = deno_core::RcRef::map(self, |r| &r.tx);
+      let mut tx = resource.borrow_mut().await;
+      let n = tx.read(&mut buf).await?;
+      Ok(n)
+    })
+  }
+
+  fn write(self: Rc<Self>, buf: ZeroCopyBuf) -> AsyncResult<usize> {
+    Box::pin(async move {
+      let resource = deno_core::RcRef::map(self, |r| &r.decoder);
+      let mut encoder = resource.borrow_mut().await;
+      let n = encoder.write(&buf).await?;
+      Ok(n)
+    })
+  }
+}
+
+pub struct DeflateCompressorResource {
+  encoder: AsyncRefCell<DeflateEncoder<DuplexStream>>,
+  tx: AsyncRefCell<DuplexStream>,
+}
+impl Resource for DeflateCompressorResource {
+  fn name(&self) -> Cow<str> {
+    "deflateCompressor".into()
+  }
+
+  fn read(self: Rc<Self>, mut buf: ZeroCopyBuf) -> AsyncResult<usize> {
+    Box::pin(async move {
+      let resource = deno_core::RcRef::map(self, |r| &r.tx);
+      let mut tx = resource.borrow_mut().await;
+      let n = tx.read(&mut buf).await?;
+      Ok(n)
+    })
+  }
+
+  fn write(self: Rc<Self>, buf: ZeroCopyBuf) -> AsyncResult<usize> {
+    Box::pin(async move {
+      let resource = deno_core::RcRef::map(self, |r| &r.encoder);
+      let mut encoder = resource.borrow_mut().await;
+      let n = encoder.write(&buf).await?;
+      Ok(n)
+    })
+  }
+}
+
+pub struct DeflateDecompressorResource {
+  decoder: AsyncRefCell<DeflateDecoder<DuplexStream>>,
+  tx: AsyncRefCell<DuplexStream>,
+}
+impl Resource for DeflateDecompressorResource {
+  fn name(&self) -> Cow<str> {
+    "deflateDecompressor".into()
+  }
+
+  fn read(self: Rc<Self>, mut buf: ZeroCopyBuf) -> AsyncResult<usize> {
+    Box::pin(async move {
+      let resource = deno_core::RcRef::map(self, |r| &r.tx);
+      let mut tx = resource.borrow_mut().await;
+      let n = tx.read(&mut buf).await?;
+      Ok(n)
```

Comment:
Shouldn't this be called the `rx` end? It is receiving, not transmitting, right?